src/sql/src/query_model/lowering.rs

```diff
@@ -0,0 +1,206 @@
+                            .iter()
+                            .map(|c| Self::lower_expression(&c.expr, &column_map))
+                            .collect_vec(),
+                    );
+                }
+
+                // Project the outer columns plus the ones in the projection of
+                // this select box
+                input.project(
+                    (0..outer_arity)
+                        .chain(input_arity..input_arity + the_box.columns.len())
+                        .collect_vec(),
+                )
+            }
+            _ => panic!(),
+        }
+    }
+
+    /// Generates a join among the result of applying the outer relation to the given
+    /// quantifiers. Returns a relation and a map of column references that can be
+    /// used to lower expressions that sit directly on top of the join.
+    fn lower_join(
+        &mut self,
+        get_outer: expr::MirRelationExpr,
+        outer_column_map: &ColumnMap,
+        quantifiers: &QuantifierSet,
+        predicates: &Vec<Box<BoxScalarExpr>>,
+    ) -> (expr::MirRelationExpr, ColumnMap) {
+        let outer_arity = get_outer.arity();
+
+        // TODO(asenac) Lower scalar subquery quantifiers
+        assert!(quantifiers.iter().all(|q_id| {
+            self.model.get_quantifier(*q_id).borrow().quantifier_type == QuantifierType::Foreach
+        }));
+
+        let join_inputs = quantifiers
+            .iter()
+            .map(|q_id| {
+                let input_box = self.model.get_quantifier(*q_id).borrow().input_box;
+                self.apply(input_box, get_outer.clone(), outer_column_map)
+            })
+            .collect_vec();
+
+        let mut join = if join_inputs.len() == 1 {
+            join_inputs.into_iter().next().unwrap()
+        } else {
+            Self::join_on_prefix(join_inputs, outer_arity)
+        };
+
+        // Generate a column map with the projection of the join plus the
+        // columns from the outer context.
+        let mut column_map = outer_column_map.clone();
+        let mut next_column = outer_arity;
+        for q_id in quantifiers.iter() {
+            let input_box = self.model.get_quantifier(*q_id).borrow().input_box;
+            let arity = self.model.get_box(input_box).borrow().columns.len();
+            for c in 0..arity {
+                column_map.insert(
+                    ColumnReference {
+                        quantifier_id: *q_id,
+                        position: c,
+                    },
+                    next_column + c,
+                );
+            }
+
+            next_column += arity;
+        }
+
+        if !predicates.is_empty() {
+            join = join.filter(
+                predicates
+                    .iter()
+                    .map(|p| Self::lower_expression(p, &column_map)),
+            );
+        }
+
+        (join, column_map)
+    }
+
+    /// Join the given inputs on a shared common prefix.
+    fn join_on_prefix(
```

Comment:
Right. There is no test that results in equivalence classes.  Only correlated queries will result in equivalence classes being generated here, since all the inputs are joined on the columns of the outer context, but right now the outer context is always the join identity. The code is prepared for supporting de-correlation, even though we don't support it yet.

Reply:
I'm ok with adding TODO comment marking sections of the code as temporarily untestable until `<specific list of additional work>` get done.

Reply:
Makes sense. Done that. Thanks!